### PR TITLE
updated with changes from mobile

### PR DESCRIPTION
**Pull request by ineiti**
While working on mobile there were some code-paths not working anymore, mostly everything
related to personhoodRPC, but also some paths through Data.

### CODE REVIEW

File: dynacred/src/Data.ts [hunk 1]
```diff
@@ -446,37 +462,14 @@ export class Data {
...
-            Log.lvl2("Registering credential");
-
-            progress("Creating Credential", 80);
-            const credentialInstance = await this.createUserCredentials(pub, darcInstances[0].id, coinInstance.id,
-                referral, contact);
-            await this.coinInstance.transfer(balance, coinInstance.id, [this.keyIdentitySigner]);
-            Log.lvl2("Registered user for darc::coin::credential:", darcInstances[0].id, coinInstance.id,
-                credentialInstance.id);
-            await contact.updateOrConnect();
+            const d = new Data(this.bc, this.storage);
```
Comment (tharvik) [inline dynacred/src/Data.ts]: Why passing `this.storage`? `Data.setValues` doesn't use any field of it.

File: dynacred/src/Data.ts [hunk 2]
```diff
@@ -752,18 +778,29 @@ export class Data {
     // createUser sets up a new user with all the necessary darcs. It does the following:
     // - creates all necessary darcs (four)
     // - creates credential and coin
-    // gData needs to have enough coins to pay for all the instances when using
+    // The user needs to have enough coins to pay for all the instances when using
     // the 'SpawnerInstance'.
     async createUser(alias: string, ephemeral?: Private): Promise<Data> {
         Log.lvl1("Starting to create user", alias);
-        const d = new Data(this.bc);
+        const d = new Data(this.bc, this.storage);
```
Comment (tharvik) [inline dynacred/src/Data.ts]: > Why passing `this.storage`? `Data.setValues` doesn't use any field of it.

Same as before.

File: dynacred/src/Data.ts [hunk 3]
```diff
@@ -660,17 +657,43 @@ export class Data {
...
+        const orgDarcs = piDarc.darc.rules.list.find((l) => l.action === "invoke:popParty.finalize").getIdentities();
+        const orgPers: Point[] = [];
+        const contacts = this.contacts.concat(this.contact);
+
+        for (const orgDarc of orgDarcs) {
+            // Remove leading "darc:" from expression
+            const orgDarcID = Buffer.from(orgDarc.substr(5), "hex");
+            const contact = contacts.find((c) => c.credentialInstance.darcID.equals(orgDarcID));
+            if (contact === undefined) {
+                return Promise.reject("didn't find organizer in contacts");
```
Comment (tharvik) [inline dynacred/src/Data.ts]: Below, you're using `throw Error(...)` to mark failure. Which one between `Promise.reject` and `throw` is the one you prefer (I lean toward the second, as converting to/from `async` means less code delta)?

File: dynacred/src/PartyItem.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
+// const ZXing = require("nativescript-zxing");
+// const QRGenerator = new ZXing();
```
Comment (tharvik) [inline dynacred/src/PartyItem.ts]: Code comments.

File: dynacred/src/personhood-rpc.ts [hunk 5]
```diff
@@ -76,69 +79,89 @@ export class PersonhoodRPC {
...
     }
 
     async wipeRPS() {
-        const ropasci = {wipe: true};
+        const ropasci = new RoPaSciList({wipe: true});
         await Promise.all(this.list.map(async (addr) => {
-            const socket = new WebSocket(addr.getWebSocketAddress(), PersonhoodRPC.serviceID);
-            // let resp = await socket.send("RoPaSciList", "RoPaSciListResponse", ropasci);
+            const socket = new WebSocketConnection(addr.getWebSocketAddress(), PersonhoodRPC.serviceID);
+            return socket.send(ropasci, RoPaSciListResponse);
```
Comment (tharvik) [inline dynacred/src/personhood-rpc.ts]: `return` and `await` would give the same results (as the return values is not read); IMHO, `await` is more readable then `return` in this case.

File: webapp/angular.json [hunk 6]
```diff
@@ -100,7 +100,8 @@
               "src/tsconfig.spec.json"
             ],
             "exclude": [
-              "**/node_modules/**"
+              "**/node_modules/**",
+	      "**/*.json"
```
Comment (tharvik) [inline webapp/angular.json]: Why did you need that? `json` files won't get build anyway if I'm correct.

File: webapp/src/app/admin/admin.component.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import { Component } from "@angular/core";
 
-import Log from "@dedis/cothority/log";
+import Log from "src/lib/cothority/log";
```
Comment (tharvik) [inline webapp/src/app/admin/admin.component.ts]: Ho yep, there is a __bunch__ of thoses lying around (52 with a rapid `grep | wc`).

File: webapp/tsconfig.json [hunk 8]
```diff
@@ -2,17 +2,18 @@
   "compileOnSave": false,
   "compilerOptions": {
     "baseUrl": "./",
+    "downlevelIteration": true,
     "outDir": "./dist/out-tsc",
     "sourceMap": true,
     "declaration": false,
-    "module": "commonjs",
+    "module": "esnext",
```
Comment (tharvik) [inline webapp/tsconfig.json]: `ESNext` is quite bleeding edge, you're sure you want that?

File: dynacred/src/Contact.ts [hunk 9]
```diff
@@ -310,6 +320,23 @@ export class Contact {
...
+                c.credentialInstance = await CredentialsInstance.fromByzcoin(bc, ul.credentialIID);
+                c.credential = c.credentialInstance.credential.copy();
+                c.darcInstance = await DarcInstance.fromByzcoin(bc, c.credentialInstance.darcID);
+            } else {
+                Log.print(ul.publicKey);
+                c.seedPublic = Public.fromProto(ul.publicKey);
+            }
+            return c;
+        } catch (e) {
+            return Log.rcatch(e, "couldn't convert to Contact from UserLocation");
```
[the remark this answers is not in the document]
Comment (ineiti) [reply dynacred/src/Contact.ts]: The point of `Log.rcatch` is to display the stack, but re-throw the error. So returning the result of `Log.rcatch` seems OK to me here.

File: dynacred/src/Contact.ts [hunk 10]
```diff
@@ -563,6 +578,15 @@ export class Contact {
                 sprintf("%s: %s", c.name, c.attributes.map((a) => a.name).join("::"))).join("\n"),
             this.recover);
     }
+
+    toUserLocation(): UserLocation {
+        return new UserLocation({
+            credential: this.credential,
+            credentialIID: this.credentialIID,
+            location: "somewhere",
```
[the remark this answers is not in the document]
Comment (ineiti) [reply dynacred/src/Contact.ts]: Yes, it's a placeholder for future functionality, and only used in `mobile`. As it's a required field in the protobuf-definition of `UserLocation`, setting it to `undefined` (or `null`) will result in failing protobuf-encoding. So I keep it as `somewhere`.

File: dynacred/src/Data.ts [hunk 11]
```diff
@@ -156,13 +156,6 @@ export class Data {
             signers: [adminSigner],
         });
 
-        try {
-            await (new OnChainSecretRPC(bc)).authorizeRoster();
```
[the remark this answers is not in the document]
Comment (ineiti) [reply dynacred/src/Data.ts]: Handled here now:

https://github.com/dedis/cothority/pull/2036/files#diff-15f18281998952f718b1ff6e6494d0a3R109

File: dynacred/src/Data.ts (hunk 1 above)
Comment (ineiti) [reply dynacred/src/Data.ts]: It's for a later `save`. But now changed to actually passing the `IStorage` class, as in a later version I'd like to pass the 'file-name' in the storage-class itself.

File: dynacred/src/Data.ts (hunk 3 above)
Comment (ineiti) [reply dynacred/src/Data.ts]: I prefer good ol' `throw`, too...

File: dynacred/src/personhood-rpc.ts (hunk 5 above)
Comment (ineiti) [reply dynacred/src/personhood-rpc.ts]: Personal opinion: I prefer `return` here, because it shows that this is the end of the processing, and that no further processing will be done. If I read `await`, I think that there will be other lines.

Here it's a real corner-case, because it's also the end of the callback-method. Anyway, `resolve conversation` or reply with `change it anyway`.

File: webapp/angular.json (hunk 6 above)
Comment (ineiti) [reply webapp/angular.json]: that's in the `lint` section. And the linter tries to lint the json-files...

File: webapp/src/app/admin/admin.component.ts (hunk 7 above)
Comment (ineiti) [reply webapp/src/app/admin/admin.component.ts]: forgot `make npm` in there...

File: dynacred/src/personhood-rpc.ts (hunk 5 above)
Comment (tharvik) [reply dynacred/src/personhood-rpc.ts]: @ineiti-bot `change it anyway`

File: webapp/tsconfig.json (hunk 8 above)
Comment (ineiti) [reply webapp/tsconfig.json]: has been done by `ng update`, reverted back to `commonjs`.

File: dynacred/src/personhood-rpc.ts (hunk 5 above)
Comment (ineiti) [reply dynacred/src/personhood-rpc.ts]: one day the bots will revolt!

File: .travis.yml [hunk 12]
```diff
@@ -14,8 +14,11 @@ jobs:
 
     - stage: build
       name: conode
+      language: go
+      go: "1.12.x"
       script:
         - cd conode
+        - go version
```
Comment (tharvik) [inline .travis.yml]: It's already available on the travis build, before running the script, in the `travis_setup_go` call.

File: dynacred/src/Contact.ts [hunk 13]
```diff
@@ -310,6 +321,23 @@ export class Contact {
...
+
+    static async fromUserLocation(bc: ByzCoinRPC, ul: UserLocation): Promise<Contact> {
+        try {
+            const c = new Contact(ul.credential);
+            if (await c.isRegisteredByzCoin(bc)) {
+                c.credentialInstance = await CredentialsInstance.fromByzcoin(bc, ul.credentialIID);
+                c.credential = c.credentialInstance.credential.copy();
+                c.darcInstance = await DarcInstance.fromByzcoin(bc, c.credentialInstance.darcID);
+            } else {
+                Log.print(ul.publicKey);
```
Comment (tharvik) [inline dynacred/src/Contact.ts]: Debug print?

File: dynacred/src/Data.ts [hunk 14]
```diff
@@ -277,32 +286,33 @@ export class Data {
         await deviceDarc.evolveDarcAndWait(newDeviceDarc, [ephemeralSigner], 5);
         return d;
     }
-    dataFileName: string;
+    dataFileName: string = "storage/data.json";
```
Comment (tharvik) [inline dynacred/src/Data.ts]: Duplicate of default with `load`.

File: dynacred/src/Defaults.ts [hunk 15]
```diff
@@ -0,0 +1,77 @@
+// import { FileIO } from "./FileIO";
```
Comment (tharvik) [inline dynacred/src/Defaults.ts]: Why is it back? I though you copied/used it in mobile, why exposing it in dynacred?

File: dynacred/src/Party.ts [hunk 16]
```diff
@@ -60,7 +60,7 @@ export class Party {
     }
 
     // qrcode(key: Public): ImageSource {
-    //     let url = Party.url + "?public=" + key.toHex();
+    //     let url = PartyItem.url + "?public=" + key.toHex();
```
Comment (tharvik) [inline dynacred/src/Party.ts]: Diff in code comment x)

File: webapp/Makefile [hunk 17]
```diff
@@ -57,6 +57,7 @@ dynacred_npm: cothority_npm
 		( echo "Moving changes to dynacred"; cp -a src/lib/dynacred/ ../dynacred/src ); \
 	rm -rf src/lib/dynacred; \
 	find src/ -name "*.ts" | xargs perl -pi -e "s:src/lib/dynacred:\@c4dt/dynacred:"
+	cd ../dynacred; npm run lint:fix
```
Comment (tharvik) [inline webapp/Makefile]: Hum, the perl changes shouldn't yield any issue with linting, why adding it?

File: webapp/angular.json [hunk 18]
```diff
@@ -100,7 +101,8 @@
               "src/tsconfig.spec.json"
             ],
             "exclude": [
-              "**/node_modules/**"
+              "**/node_modules/**",
+              "**/*.json"
```
Comment (tharvik) [inline webapp/angular.json]: As `build-angular:tslint` respects `tsconfig.json`, there is no need to exclude things (especially JSON ones)

File: webapp/src/app/user-data.service.ts [hunk 19]
```diff
@@ -29,6 +31,19 @@ export class UserData extends Data {
         const config = Config.fromTOML(await res.text());
         this.bc = await ByzCoinRPC.fromByzcoin(config.roster, config.byzCoinID);
     }
+
+    async load() {
```
Comment (tharvik) [inline webapp/src/app/user-data.service.ts]: What's the use? `d` is not returned

File: dynacred/package.json [hunk 20]
```diff
@@ -11,8 +11,10 @@
   "scripts": {
     "build": "tsc",
     "test": "ts-node --prefer-ts-exts node_modules/.bin/jasmine",
-    "prepare": "tslint --project . && npm run build",
-    "cover": "nyc jasmine"
+    "prepare": "npm run lint && npm run build",
+    "cover": "nyc jasmine",
+    "lint": "tslint --project .",
+    "lint:fix": "tslint --fix --project ."
```
[the remark this answers is not in the document]
Comment (ineiti) [reply dynacred/package.json]: I prefer `cat package.json` to `npm run lint --help` whenever I forget all the possible arguments. So the `package.json` is kind of my library of useful things to do...

File: dynacred/src/Badge.ts [hunk 21]
```diff
@@ -1,35 +1,27 @@
...
+import CoinInstance from "@dedis/cothority/byzcoin/contracts/coin-instance";
+import SpawnerInstance from "@dedis/cothority/personhood/spawner-instance";
 import { Contact } from "./Contact";
 import { Data } from "./Data";
 import { KeyPair } from "./KeyPair";
-import { Party } from "./Party";
+import { PartyItem } from "./PartyItem";
 
 // const ZXing = require("nativescript-zxing");
 // const QRGenerator = new ZXing();
```
[the remark this answers is not in the document]
Comment (ineiti) [reply dynacred/src/Badge.ts]: how did get that one back?

File: .travis.yml (hunk 12 above)
Comment (ineiti) [reply .travis.yml]: yes, sorry - debugging stuff leftover...

File: dynacred/src/Party.ts (hunk 16 above)
Comment (ineiti) [reply dynacred/src/Party.ts]: leftover file - deleted.

File: webapp/Makefile (hunk 17 above)
Comment (ineiti) [reply webapp/Makefile]: yes, linting changes, because the order of the imports is not correct anymore with respect to the new paths...

File: webapp/angular.json (hunk 18 above)
Comment (ineiti) [reply webapp/angular.json]: `src/lib/cothority/protobuf/models.json` gives errors in linting. Will disappear if I get around to replace `make src` with `npm link ../dynacred && inotify...`

File: webapp/src/app/user-data.service.ts (hunk 19 above)
Comment (ineiti) [reply webapp/src/app/user-data.service.ts]: That's the old style loading that is still needed with the data-service, as you load it before the system has the possibility of fetching the configuration.
